### PR TITLE
Support for 10x HDF5 input data; Input data validation and sanitation

### PR DESCRIPTION
**Pull request by fedorgrab**

### CODE REVIEW

File: src/cas_cli/_read_data.py [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+            raise Exception("File is missing one or more required datasets.")
+
+
+def read_10x_h5(
+    filename: t.Union[str, "pathlib.Path"],
+    genome: t.Optional[str] = None,
+) -> "anndata.AnnData":
+    """
+     Read 10x-Genomics-formatted hdf5 file.
+     Refer to scanpy library https://github.com/scverse/scanpy/blob/1.6.x/scanpy/readwrite.py#LL138C10-L138C10.
```
Comment (mbabadi) [inline src/cas_cli/_read_data.py:112]: This method is borrowed from scanpy, please refer to ...

File: src/cas_cli/client.py [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+from cas_cli import _read_data, data_preparation, exceptions, service
+
+
+class CASClient:
+    """
+    Service that is designed to call Cellarium Cloud Backend API service
+    """
+
+    def __init__(self, api_token: str) -> None:
+        self._print("Connecting with Cellarium Cloud backend...")
```
Comment (mbabadi) [inline src/cas_cli/client.py:21]: with => to

File: src/cas_cli/client.py [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+        try:
+            data_preparation.validate(
+                adata=adata,
+                cas_feature_schema_list=cas_feature_schema_list,
+                feature_ids_column_name=feature_ids_column_name,
+            )
+        except exceptions.DataValidationError as e:
+            if e.extra_features > 0:
+                self._print(
+                    f"Input datafile has {e.extra_features} more features than {feature_schema_name} schema in CAS. "
```
Comment (mbabadi) [inline src/cas_cli/client.py:59]: f'The input data matrix has {e.extra_features} extra features compared to "{feature_schema_name}" CAS schema ({len(cas_feature_schema_list)}).'

File: src/cas_cli/client.py [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+                    f"They will be omitted..."
```
Comment (mbabadi) [inline src/cas_cli/client.py:60]: "They will be omitted..." => "Extra input features will be dropped."

File: src/cas_cli/client.py [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+                )
+            if e.missing_features > 0:
+                self._print(
+                    f"Input datafile has {e.missing_features} missing features compared with {feature_schema_name} in CAS. "
```
Comment (mbabadi) [inline src/cas_cli/client.py:64]: f'The input data matrix has {e.missing_features} missing features compared to "{feature_schema_name}" CAS schema ({len(cas_feature_schema_list)}).'

File: src/cas_cli/client.py [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+                    f"They will be filled with zeros..."
```
Comment (mbabadi) [inline src/cas_cli/client.py:65]: Missing features will be imputed with zeros.

File: src/cas_cli/client.py [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+                )
+            if e.extra_features == 0 and e.missing_features == 0:
+                self._print(
+                    f"Input datafile has all the necessary features as {feature_schema_name}, but it's still "
```
Comment (mbabadi) [inline src/cas_cli/client.py:69]: f'The input data matrix contains all of the features specified in "{feature_schema_name}" CAS schema but '
f'in a different order. The input features will be reordered according to "{feature_schema_name}".'

File: src/cas_cli/client.py [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+                    f"according to {feature_schema_name}..."
+                )
+            return data_preparation.sanitize(
+                adata=adata,
+                cas_feature_schema_list=cas_feature_schema_list,
+                count_matrix_name=count_matrix_name,
+                feature_ids_column_name=feature_ids_column_name,
+            )
+        else:
+            self._print(f"Input dataset is validated and fully compatible with {feature_schema_name} schema...")
```
Comment (mbabadi) [inline src/cas_cli/client.py:80]: f'The input data matrix conforms with the "{feature_schema_name}" CAS schema.'

File: src/cas_cli/client.py [hunk 9]
```diff
@@ -0,0 +1,238 @@
...
+        In case of HTTP 501 print a message
+
+        :param adata_bytes: Dumped bytes of anndata chunk
+        :param results: Results list that needs to be used to inplace the response from the server
+        :param chunk_index: Consequent number of the chunk (e.g Chunk 1, Chunk 2)
+        :param chunk_start_i: Index pointing to the main adata file start position of the current chunk
+        :param chunk_end_i: Index pointing to the main adata file end position of the current chunk
+        """
+        number_of_cells = chunk_end_i - chunk_start_i
+        for _ in range(3):
```
Comment (mbabadi) [inline src/cas_cli/client.py:103]: 3 is a magic number. Replace with `num_attempts_per_chunk` and expose to the user.

Comment (mbabadi) [reply src/cas_cli/client.py:103]: make it a kwarg of `CASClient.__init__` with a default value of 3.

File: src/cas_cli/client.py [hunk 10]
```diff
@@ -0,0 +1,238 @@
...
+            try:
+                results[chunk_index] = await self.cas_api_service.async_annotate_anndata_chunk(
+                    adata_file_bytes=adata_bytes, number_of_cells=number_of_cells
+                )
+
+            except exceptions.HTTPError500:
+                self._print(
+                    f"Something went wrong, resubmitting chunk #{chunk_index + 1:2.0f} ({chunk_start_i:5.0f}, {chunk_end_i:5.0f}) to CAS ..."
```
Comment (mbabadi) [inline src/cas_cli/client.py:111]: "Something went wrong" => "Error occurred in CAS backend (HTTPError500)"

File: src/cas_cli/client.py [hunk 11]
```diff
@@ -0,0 +1,238 @@
...
+                    chunk_index=chunk_index,
+                    chunk_start_i=chunk_start_i,
+                    chunk_end_i=chunk_end_i,
+                )
+            )
+            i = j
+            j += chunk_size
+
+        await asyncio.wait(tasks)
+
```
Comment (mbabadi) [inline src/cas_cli/client.py:161]: make an appropriate `annotate_anndata_file` method. use `AnnData.read_h5ad`.

File: src/cas_cli/client.py [hunk 12]
```diff
@@ -0,0 +1,238 @@
...
+    def annotate_anndata(
+        self,
+        adata: "anndata.AnnData",
+        chunk_size=2000,
+        feature_schema_name: str = "default",
+        count_matrix_name: str = "X",
+        feature_ids_column_name: str = "index",
+    ) -> t.List[t.Dict[str, t.Any]]:
+        """
+        Send an anndata object to Cellarium Cloud backend to get annotations. Split the input
```
Comment (mbabadi) [inline src/cas_cli/client.py:171]: here and everywhere in the docstrings, please replace `anndata` with `anndata.AnnData`.

File: src/cas_cli/data_preparation.py [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+import numpy as np
+import pandas as pd
+import scipy.sparse as sp
+
+from cas_cli import exceptions
+
+
+def validate(
+    adata: anndata.AnnData,
+    cas_feature_schema_list: t.List,
```
Comment (mbabadi) [inline src/cas_cli/data_preparation.py:14]: Make an issue for generalization and unification of the concept of CAS schema for the 3 use cases:
- reference data preparation and training curriculum generation
- client-side data validation and sanitation
- pytorch dataloader

File: src/cas_cli/data_preparation.py [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+
+def sanitize(
+    adata: anndata.AnnData,
+    cas_feature_schema_list: t.List[str],
+    count_matrix_name: str,
+    feature_ids_column_name: str,
+) -> anndata.AnnData:
+    """
+    Cellarium CAS sanitizing script. Returns a new `anndata.AnnData` instance, based on the feature expression
+    matrix of the input instance. Extra Features get omitted. Missing features get filled with zeros.
```
Comment (mbabadi) [inline src/cas_cli/data_preparation.py:56]: Extra Feature => Extra features

File: src/cas_cli/data_preparation.py [hunk 15]
```diff
@@ -0,0 +1,105 @@
...
+    feature_id_intersection_adata_indices = list(map(adata_feature_id_map.get, feature_id_intersection))
+
+    with warnings.catch_warnings():
+        warnings.simplefilter("ignore", sp.SparseEfficiencyWarning)
+        n_cells = adata.shape[0]
+        n_features = len(cas_feature_schema_list)
+        result_matrix = sp.csc_matrix((n_cells, n_features), dtype=np.float32)
+        input_matrix = adata.X if count_matrix_name == "X" else adata.raw.X
+
+        result_matrix[:, feature_id_intersection_cas_indices] = input_matrix.tocsc()[
```
Comment (mbabadi) [inline src/cas_cli/data_preparation.py:92]: (low priority) make an enhancement issue for optimizing input anndata chunking and reformatting operations. Slicing CSR matrices is cheap. We can slice off the chunk that we need and only convert that slice to CSC and then column slicing by `feature_id_intersection_cas_indices`.

File: tests/unit/test_data_preparation.py [hunk 16]
```diff
@@ -0,0 +1,276 @@
...
+            "ENSG00000253811",
+            "ENSG00000186439",
+            "ENSG00000254756",
+            "ENSG00000139211",
+            "ENSG00000185615",
+            "ENSG00000246379",
+        ]
+
+    @staticmethod
+    def get_bigger_feature_schema() -> t.List:
```
Comment (mbabadi) [inline tests/unit/test_data_preparation.py:53]: bigger => larger

File: tests/unit/test_data_preparation.py [hunk 17]
```diff
@@ -0,0 +1,276 @@
...
+            "ENSG00000287710",
+        ]
+
+    def test_data_validation_input_has_missing_features(self):
+        cas_feature_schema_list = self.get_feature_schema()
+        adata_feature_schema_list = self.get_smaller_feature_schema()
+        data_validation_error_was_raised = False
+        number_of_missing_features = 0
+        number_of_extra_features = 0
+        d = np.random.randint(0, 500, size=(self.INPUT_DATASET_LENGTH, len(adata_feature_schema_list)))
```
Comment (mbabadi) [inline tests/unit/test_data_preparation.py:114]: It's always best to use private seeded random number generators (`np.random.RandomState(seed=...)`) in unit tests rather than the global `np.random`.

**Review by mbabadi — CHANGES_REQUESTED** [review]
Nicely done! :)

File: src/cas_cli/data_preparation.py (hunk 15 above)
Comment (fedorgrab) [reply src/cas_cli/data_preparation.py:92]: https://github.com/cellarium-ai/cell-annotation-service-client/issues/14

File: src/cas_cli/data_preparation.py (hunk 13 above)
Comment (fedorgrab) [reply src/cas_cli/data_preparation.py:14]: https://github.com/cellarium-ai/cellarium-cloud/issues/80

File: src/cas_cli/client.py [hunk 18]
```diff
@@ -56,28 +59,35 @@ def _validate_and_sanitize_input_data(
         except exceptions.DataValidationError as e:
             if e.extra_features > 0:
                 self._print(
-                    f"Input datafile has {e.extra_features} more features than {feature_schema_name} schema in CAS. "
-                    f"They will be omitted..."
+                    f"The input data matrix has {e.extra_features} extra features compared to '{feature_schema_name}' "
+                    f"CAS schema ({len(cas_feature_schema_list)})."
```
Comment (mbabadi) [inline src/cas_cli/client.py:63]: Add a whitespace after `.`.

File: src/cas_cli/client.py [hunk 19]
```diff
@@ -56,28 +59,35 @@ def _validate_and_sanitize_input_data(
...
+                    f"Extra input features will be dropped."
                 )
             if e.missing_features > 0:
                 self._print(
-                    f"Input datafile has {e.missing_features} missing features compared with {feature_schema_name} in CAS. "
-                    f"They will be filled with zeros..."
+                    f"The input data matrix has {e.missing_features} missing features compared to "
+                    f"'{feature_schema_name}' CAS schema ({len(cas_feature_schema_list)}). \n"
```
Comment (mbabadi) [inline src/cas_cli/client.py:69]: Remove `\n` after the whitespace.

File: src/cas_cli/client.py [hunk 20]
```diff
@@ -209,7 +220,35 @@ def annotate_anndata(
...
+    def annotate_anndata_file(
+        self,
+        filepath: str,
+        chunk_size=2000,
+        feature_schema_name: str = "default",
+        count_matrix_name: str = "X",
+        feature_ids_column_name: str = "index",
+    ) -> t.List[t.Dict[str, t.Any]]:
+        """
+        Read 10x `anndata.AnnData` matrix and apply the annotate method on it.
```
Comment (mbabadi) [inline src/cas_cli/client.py:232]: It's not 10x `anndata.AnnData`, it's just `anndata.AnnData` :D

**Review by mbabadi — CHANGES_REQUESTED** [review]
a few small things

File: src/cas_cli/client.py (hunk 20 above)
Comment (fedorgrab) [reply src/cas_cli/client.py:232]: Omg, right...

**Comment by mbabadi** [comment]
@fedorgrab while you're at it, please update README for different use cases (10x HDF5 file, AnnData file, loaded AnnData object).
